monitor/supply/supply.go

```diff
@@ -0,0 +1,93 @@
+package supply
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/omni-network/omni/contracts/bindings"
+	"github.com/omni-network/omni/lib/cchain"
+	"github.com/omni-network/omni/lib/contracts"
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/ethclient"
+	"github.com/omni-network/omni/lib/log"
+	"github.com/omni-network/omni/lib/netconf"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/params"
+
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+func MonitorForever(ctx context.Context, cprov cchain.Provider, network netconf.Network, ethCls map[uint64]ethclient.Client) {
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			if err := instrSupplies(ctx, cprov, network, ethCls); err != nil {
+				log.Warn(ctx, "Token supply intrumentation failed", err)
+			}
+
+			timer.Reset(time.Hour)
+		}
+	}
+}
+
+// instrSupplies instruments the supplies of OMNI token on the L1 (without the bridge balances)
+// and on the consensus chain.
+func instrSupplies(ctx context.Context, cprov cchain.Provider, network netconf.Network, ethCls map[uint64]ethclient.Client) error {
+	response, err := cprov.QueryClients().Bank.TotalSupply(ctx, &types.QueryTotalSupplyRequest{})
+	if err != nil {
+		return errors.Wrap(err, "total supply query")
+	}
+	cosmosSupplyWei := math.NewInt(0)
+	for _, coin := range response.Supply {
+		if coin.Denom == sdk.DefaultBondDenom {
+			cosmosSupplyWei = cosmosSupplyWei.Add(coin.Amount)
+		}
+	}
```

Comment:
nit: although this works, I'd reads like "supply" may contain multiple "bond" denoms which we need to sum, which isn't correct.

```
func stakeAmount(coins sdk.Coins) *big.Int {
	if len(coins) != 1 {
		return errors.new("unexpected number of coins")
	}
	ok, coin := resp.Supply.Find(sdk.DefaultBondDenom)
	if !ok {
		return errors.new("missing default bond denom")
	}
	return coin.Amount.BigInt()
}
```